src/zkey/ep11.c

```diff
@@ -402,3 +446,421 @@ int reencipher_ep11_key(struct ep11_lib *ep11, target_t target,
+		{ CKA_IBM_PROTKEY_EXTRACTABLE, &cktrue, sizeof(CK_BBOOL) },
+		{ CKA_SIGN, &cktrue, sizeof(CK_BBOOL) },
+	};
+	struct ep11kblob_header *ep11hdr;
+
+	if (ep11->dll_m_EncryptSingle == NULL || ep11->dll_m_UnwrapKey == NULL) {
+		DEBUG("EP11 host lib function m_EncryptSingle and/or m_UnwrapKey "
+			"not available but required for EP11 EC clear key import.");
+		return -ELIBACC;
+	}
+
+	/* Create AES wrapping key */
+	rv = make_wrapblob(ep11, target);
+	if (rv != CKR_OK) {
+		DEBUG("Could not create the wrap blob");
+		goto ret;
+	}
+
+	/* Create ASN.1 key struct for m_EncryptSingle */
+	make_asn1_key_sequence(curve, pubkey, publen, privkey, privlen,
+						(CK_BYTE *)&asn1_buf, &asn1_len);
+
+	/* Encrypt EC private key with AES wrapping key */
+	rv = ep11->dll_m_EncryptSingle(ep11->raw2key_wrap_blob,
+						ep11->raw2key_wrap_blob_l, &mech_w,
+						(unsigned char *)asn1_buf, asn1_len,
+						cipher, &cipher_len, target);
+	if (rv != CKR_OK) {
+		DEBUG("Error m_EncryptSingle, rv = 0x%lx", rv);
+		goto ret;
+	}
+
+	/* Add default attributes */
+	memcpy(attrs, &default_attrs, sizeof(default_attrs));
+	attrs_len = sizeof(default_attrs) / sizeof(CK_ATTRIBUTE);
+
+	/* Add user-defined flags, translated to attributes */
+	ec_key_add_attrs_from_flags(flags, attrs, &attrs_len, &cktrue);
+
+	/* Create the secure key blob */
+	rv = ep11->dll_m_UnwrapKey(cipher, cipher_len, ep11->raw2key_wrap_blob,
+			ep11->raw2key_wrap_blob_l, NULL, ~0, ep11_pin_blob,
+			ep11_pin_blob_len, &mech_w, (CK_ATTRIBUTE *)&attrs, attrs_len,
+			blob, &blob_len, csum, &cslen, target);
+	if (rv != CKR_OK) {
+		DEBUG("Error m_UnwrapKey, rv = 0x%lx", rv);
+		goto ret;
+	}
+
+	/* Add info to ep11 header (overlay over session ID field). This info
+	 * is later needed for reencipher. */
+	ep11hdr = (struct ep11kblob_header *)&blob;
+	ep11hdr->len = blob_len;
+	ep11hdr->version = PKEY_TYPE_EP11_ECC;
+	ep11hdr->bitlen = curve2bitlen[curve];
+
+	/* Copy result */
+	memcpy(secure_key, blob, blob_len);
+	*seclen = blob_len;
+
+	rv = CKR_OK;
+
+ret:
+
+	switch (rv) {
+	case CKR_OK:
+		rc = 0;
+		break;
+	default:
+		rc = -EIO;
+		break;
+	}
+
+	return rc;
+}
+
+/**
+ * Extract the public key from given secure ECC key token. To obtain a public
+ * key here, it must be contained in the token. The host lib does not
+ * re-calculate a public key from given secure key.
+ */
+int ec_key_extract_public_ep11(struct ep11_lib *ep11, unsigned char *secure_key,
+				unsigned int seclen, unsigned char *public_key, unsigned int *publen,
+				target_t target)
+{
+	CK_RV rv;
+	CK_BYTE point[132];
+	size_t point_len = sizeof(point);
+	CK_ATTRIBUTE templ[] = {
+		{ CKA_EC_POINT, &point, sizeof(point) },
+	};
+	int rc = -EIO;
+
+	if (ep11->dll_m_GetAttributeValue == NULL) {
+		DEBUG("EP11 host lib function m_GetAttributeValue not "
+			"available but required for EP11 EC extract public key.");
+		return -ELIBACC;
+	}
+
+	rv = ep11->dll_m_GetAttributeValue(secure_key, seclen, templ, 1, target);
```

Comment:
Code is reworked: all keys are now "keys with header" and the call looks like:
```
/* Note that the secure key is a TOKVER_EP11_ECC_WITH_HEADER and has a
 * 16-byte ep11kblob_header prepended before the actual secure key blob.
 * This header must not be passed to the ep11 library. */
rv = ep11->dll_m_GetAttributeValue(secure_key + sizeof(struct ep11kblob_header),
					seclen - sizeof(struct ep11kblob_header),
					templ, 1, target);
```
For the ep11 lib call I just have to skip the header.

Reply:
Right, the call looks good now.

I was wondering if the EP11 host lib supports a GetAttribueValues with  CKA_EC_POINT at all. Its not documented that it would support so (and I can't see that in the EP11 code being implemented either). So I assume that this call will actually never work.....

Newer host libs support GetAttribueValues with CKA_PUBLIC_KEY_INFO to get the SPKI....

Reply:
Good catch. Reworked to use CKA_IBM_MACED_PUBLIC_KEY_INFO, which however is only supported with ep11 hostlib v4.